rfcs/011-opentracing.md

```diff
@@ -0,0 +1,233 @@
+document are to be interpreted as described in
+[RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.txt).
+
+## Terminology
+
+- [span](https://github.com/opentracing/specification/blob/1.1/specification.md#the-opentracing-data-model): The "span"
+  is the primary building block of a distributed trace, representing an individual unit of work done in a distributed system.
+  Each Span encapsulates the following state:
+   - An operation name
+   - A start timestamp
+   - A finish timestamp
+   - A set of zero or more key:value `Span Tags`. 
+   - A set of zero or more structured logs (key:value `Span Logs`).
+   - A `SpanContext`
+   - `References` to zero or more causally-related `Spans`
+
+---
+
+# Detailed Description
+
+## Selection of a library
+
+As mentioned earlier, there are two flavours of libraries. None of them is perfect for all use cases.
+The biggest differences in between `otters` and `passage` are:
+
+|                                | otters      | passage                   |
+| ------------------------------ | ----------- | ------------------------- |
+| reporting protocol             | http        | udp                       |
+| filtering                      | custom DSL  | sampling callback module  |
+| reporter                       | zipkin only | jaeger or plugin          |
+| functional API                 |      +      |             +             |
+| process dictionary             |      +      |             +             |
+| process based span storage     |      +      |             -             |
+| send event in batches          |      +      |             -             |
+| sender overload detection      |      -      |             +             |
+| report batches based on        | timer       | spans of single operation |
+| design for performance         |      +      |             -             |
+| design for robustness at scale |      -      |             +             |
+| counters                       |      +      |             -             |
+| sampling based on duration     |      +      |             -             |
+
+In order to allow future replacement of a tracing library it would be desirable to create an interface module `couch_trace`.
+The `otters` library would be used for the first iteration.
+
+## Configuration
+
+The `otters` library uses application environment to store its configuration. 
+It also has a facility to compile filtering DSL into a beam module.
+The filtering DSL looks like following: `<name>([<condition>]) -> <action>.`. 
+The safety of DSL compiler is unknown. Therefore a modification of tracing settings via configuration over HTTP wouldn't be possible.
+The otter related section of the config `otter.filters` would be protected by BLACKLIST_CONFIG_SECTIONS.
+The configuration of tracing would only be allowed from remsh or modification of the ini file.
+The configuration for otter filters would be stored in couch_config as follows:
+```
+[otter.filters]
```

Comment:
After writing documentation for "ken" and "smoosh" ... I wonder if we should rather call this section something like `tracing.filters` to avoid yet another explanation for why a config section has such an obfuscated name :)